packages/@orbit/jsonapi/src/jsonapi-request-processor.ts

```diff
@@ -0,0 +1,256 @@
+      serializer: this.serializer,
+      keyMap: settings.keyMap
+    });
+    this.initDefaultFetchSettings(settings);
+  }
+
+  fetch(url: string, customSettings?: FetchSettings): Promise<any> {
+    let settings = this.initFetchSettings(customSettings);
+
+    let fullUrl = url;
+    if (settings.params) {
+      fullUrl = this.urlBuilder.appendQueryParams(fullUrl, settings.params);
+      delete settings.params;
+    }
+
+    let fetchFn = (Orbit as any).fetch || Orbit.globals.fetch;
+
+    // console.log('fetch', fullUrl, settings, 'polyfill', fetchFn.polyfill);
+
+    if (settings.timeout) {
+      let timeout = settings.timeout;
+      delete settings.timeout;
+
+      return new Promise((resolve, reject) => {
+        let timedOut: boolean;
+
+        let timer = Orbit.globals.setTimeout(() => {
+          timedOut = true;
+          reject(new NetworkError(`No fetch response within ${timeout}ms.`));
+        }, timeout);
+
+        fetchFn(fullUrl, settings)
+          .catch((e: Error) => {
+            Orbit.globals.clearTimeout(timer);
+
+            if (!timedOut) {
+              return this.handleFetchError(e);
+            }
+          })
+          .then((response: any) => {
+            Orbit.globals.clearTimeout(timer);
+
+            if (!timedOut) {
+              return this.handleFetchResponse(response);
+            }
+          })
+          .then(resolve, reject);
+      });
+    } else {
+      return fetchFn(fullUrl, settings)
+        .catch((e: Error) => this.handleFetchError(e))
+        .then((response: any) => this.handleFetchResponse(response));
+    }
+  }
+
+  initFetchSettings(customSettings: FetchSettings = {}): FetchSettings {
+    let settings: FetchSettings = deepMerge({}, this.defaultFetchSettings, customSettings);
+
+    if (settings.json) {
+      assert('`json` and `body` can\'t both be set for fetch requests.', !settings.body);
+      settings.body = JSON.stringify(settings.json);
+      delete settings.json;
+    }
+
+    if (settings.headers && !settings.body) {
+      delete (settings.headers as any)['Content-Type'];
+    }
+
+    return settings;
+  }
+
+  operationsFromDeserializedDocument(deserialized: RecordDocument): Operation[] {
+    const records: Record[] = [];
+    Array.prototype.push.apply(records, toArray(deserialized.data));
+
+    if (deserialized.included) {
+      Array.prototype.push.apply(records, deserialized.included);
+    }
+
+    return records.map(record => {
+      return {
+        op: 'updateRecord',
+        record
+      };
+    });
+  }
+
+  buildFetchSettings(options: RequestOptions = {}, customSettings?: FetchSettings): FetchSettings {
+    return buildFetchSettings(options, customSettings);
+  }
+
+  customRequestOptions(queryOrTransform: Query | Transform): RequestOptions {
+    return deepGet(queryOrTransform, ['options', 'sources', this.sourceName]);
+  }
+
+  processMeta(meta:any, queryOrRequest:Query|TransformRecordRequest) {}
```

Comment:
Perhaps this handler shouldn't be so targeted at processing `meta` but should instead allow for a pre-processing step on response documents in general?

In other words, it could be refactored:

```
preprocessResponseDocument(document: ResourceDocument, queryOrTransform: Query | Transform) {}
```

This callback could be used to extract `meta`, `links`, or even to massage responses prior to deserialization.

Reply:
I like making this more expansive.

It seems like it begins to overlap with `handleFetchResponse` though. The big difference  is having access to the query or transformRequest that resulted in the response which I've found is critical context to deciding how to process the data.

Finally, I'm not sure how to think about calling this before vs. after deserialization. I guess before makes the most sense.

Reply:
> The big difference is having access to the query or transformRequest that resulted in the response which I've found is critical context to deciding how to process the data.

Yeah, I suspect that's key for making sense of the response document. Also, `handleFetchResponse` is operating at a lower-level and deals with responses that may or may not have content.

> Finally, I'm not sure how to think about calling this before vs. after deserialization. I guess before makes the most sense. 

We could also add `postprocessResponseDocument` that would take a deserialized document of type `RecordDocument`.

Reply:
Sounds good. We can start with `preprocessResponseDocument` and we'll see if we need the `postprocess...` method as we start using it.

Reply:
Ok, SGTM too 👍